projects/components/src/download-file/service/file-download.service.ts

```diff
@@ -31,25 +31,17 @@ export class FileDownloadService {
    * @param config Csv download config
    */
   public downloadAsCsv(config: CsvDownloadFileConfig): Observable<FileDownloadEvent> {
-    // If given header is empty, then create header from the data keys
-    const header$ = isEmpty(config.header)
-      ? config.dataSource.pipe(
-          map(data => Object.keys(data[0] ?? [])),
-          map(keys => keys.map(startCase))
-        )
-      : of(config.header!);
-
+    const getAllDataKeys = (data: Dictionary<unknown>[]): string[] => uniq(data.map(row => Object.keys(row)).flat());
     // Value replacer for null and undefined values
     const replacer = (_: string, value: string) => value ?? '-';
+    const csvData$ = config.dataSource.pipe(
+      map(data => {
+        const allKeys = isEmpty(config.header) ? getAllDataKeys(data) : config.header!;
```

Comment:
One major concern here is there's a big assumption that the header provided is keys, and not already cased/renamed. Do we know that to be true for all use cases? If it is, I wouldn't even call it the header any more since that sounds like it should not impact the extraction. Would call it like csvColumn with a key and optional title or something.